src/biota_models/vegetation/simulation/veg_simulation_2species.py

```diff
@@ -0,0 +1,410 @@
+from src.core.simulation.base_simulation import BaseSimulation
+
+
+class _VegetationSimulation_2species(BaseSimulation, ABC):
+    """
+    Implements the `SimulationProtocol`.
+    Facade class that can be implemented through an Adapter pattern.
+    VegetationModel simulation.
+    """
+
+    # Other fields.
+    constants: Optional[VegetationConstants]
+    output: Optional[VegOutputWrapper]
+    output2: Optional[VegOutputWrapper]
+    veg: Optional[Vegetation]
+    veg2: Optional[Vegetation]
+
+    @validator("constants", pre=True, allow_reuse=True)
+    @classmethod
+    def validate_constants(
+        cls, field_value: Union[str, Path, VegetationConstants]
+    ) -> VegetationConstants:
+        """
+        Validates the user-input constants value and transforms in case it's a filepath (str, Path).
+
+        Args:
+            field_value (Union[str, Path, Constants]): Value given by the user representing Constants.
+
+        Raises:
+            NotImplementedError: When the input value does not have any converter.
+
+        Returns:
+            Constants: Validated constants value.
+        """
+        if isinstance(field_value, VegetationConstants):
+            return field_value
+        if isinstance(field_value, str):
+            field_value = Path(field_value)
+        if isinstance(field_value, Path):
+            return VegetationConstants.from_input_file(field_value)
+        raise NotImplementedError(f"Validator not available for {type(field_value)}")
+
+    @validator("veg", pre=True, allow_reuse=True)
+    @classmethod
+    def validate_vegetation(
+        cls, field_value: Union[dict, Vegetation], values: dict
+    ) -> Vegetation:
+        """
+        Initializes vegetation in case a dictionary is provided. Ensuring the constants are also
+        given to the object.
+
+        Args:
+            field_value (Union[dict, Vegetation]): Value given by the user for the Vegetation field.
+            values (dict): Dictionary of remaining user-given field values.
+
+        Returns:
+            Vegetation: Validated instance of 'Vegetation'.
+        """
+        if isinstance(field_value, Vegetation):
+            return field_value
+        if isinstance(field_value, dict):
+            # Check if constants present in the dictionary:
+            if "constants" in field_value.keys():
+                # It will be generated automatically.
+                # in case parameters are missing an error will also be displayed.
+                return Vegetation(**field_value)
+            if "constants" in values.keys():
+                field_value["constants"] = values["constants"]
+                return Vegetation(**field_value)
+            raise ValueError(
+                "Constants should be provided to initialize a Vegetation Model."
+            )
+        raise NotImplementedError(f"Validator not available for {type(field_value)}")
+
+    @validator("veg2", pre=True, allow_reuse=True)
```

Comment:
is this doing the same as the 'veg' validator? if so you can just do @validator(["veg", "veg2"], pre=True, allow_reuse=True)
Same applies for output
https://pydantic-docs.helpmanual.io/usage/validators/